### PR TITLE
Add support for pausing / resuming windups

### PR DESCRIPTION
**Pull request by anulman**
Hello! I was trying to pause / resume my windup based on interactions with footnotes embedded in the text (see Loom). Thought I would contrib my work so far.

I hear you're working on a new rendering mechanism (backwards and forwards! so cool!!) so no worries if this shan't land. If you're into it please advise re: how to test and/or document this for release.

https://www.loom.com/share/d0e2f1c980d44bdc9bd49d15208d676c

### CODE REVIEW

**Comment by anulman** [comment]
I'd love try the alpha! Meanwhile it sounds like the API will involve manually adjusting the children to the "paused" / "resumed" values? If so I'd _much_ prefer a simple `isPaused`-like API; it wouldn't really work for [my use case](https://github.com/anulman/bigcomputer.xyz/blob/main/src/pages/index.tsx#L82).

To be clear, I'm very much not complaining; I trialed all the "typey text" libs I could find and _love_ windups; I'd be happy to help ship any adjustments if you'd point me to the source!

File: src/react/useWindup.ts [hunk 1]
```diff
@@ -100,6 +106,36 @@ export default function useWindup<M extends HookMetadata>(
...
+    if (timeoutRef.current) {
+      isPausedRef.current = true;
+
+      clearTimeout(timeoutRef.current);
+      pauseDelayRemainingRef.current = Math.max(
+        0,
+        nextCharAtRef.current ?? 0 - Date.now()
+      );
+    }
+  }, [isPausedRef, timeoutRef, nextCharAtRef, pauseDelayRemainingRef]);
```
Comment (sgwilym) [inline src/react/useWindup.ts:123]: You can safely leave the refs out of the dependency array.

File: src/react/useWindup.ts [hunk 2]
```diff
@@ -100,6 +106,36 @@ export default function useWindup<M extends HookMetadata>(
...
+    }
+
+    if (!windupIsFinished) {
+      isPausedRef.current = false;
+
+      setTimeout(() => {
+        dispatch({ type: "next" });
+      }, pauseDelayRemainingRef.current ?? 0);
+    }
+  }, [isPausedRef, windupIsFinished, pauseDelayRemainingRef]);
```
Comment (sgwilym) [inline src/react/useWindup.ts:137]: Same here re: refs.

**Review by sgwilym — CHANGES_REQUESTED** [review]

**Comment by sgwilym** [comment]
@anulman I've re-read everything again, and I've misunderstood. That rewindup API I suggested doesn't really help you at all! I get the usecase and its context now. I'd be happy to add this, just a few thoughts.

I do wonder if this easy to confuse with the `<Pause>` component and what that does, given there are a few names that span the hook and component APIs (e.g. the `pace` option and `<Pace>`). Pausing seems like the right name for the feature you're introducing, perhaps the current `<Pause>` should become `<Wait>`?

**Comment by sgwilym** [comment]
You know, I completely misunderstood this. And yeah, the Rewindup API doesn't help you at all!

Now that I understand the context, I see what this API is for. Let's add it! A few thoughts.

There are a few names that span both the hook and component APIs (e.g. `pace` and `<Pace>`). I wonder if this new API could be conflated with the `<Pause>` API.  Pausing is a pretty appropriate name for what you're doing, maybe `<Pause>` could become `<Wait>`?

File: src/react/useWindup.ts (hunk 1 above)
Comment (anulman) [reply src/react/useWindup.ts:123]: Oop, good point! Will do

**Comment by anulman** [comment]
Yeah, I like the idea of renaming `<Pause>` to `<Wait>`, however I didn't want to introduce breaking changes! What about:

- Rename `<Pause>` to `<Wait>`
- Create a `<Pause>` component, which proxies to `<Wait>`
- Mark the `<Pause>` component as deprecated

Also—I fixed the subtle timing issue; see the above commit :)

**Review by sgwilym — APPROVED** [review]

**Comment by sgwilym** [comment]
Thank you for this addition, and your patience! I'll introduce the Pause/Wait stuff in another change.

**Comment by sgwilym** [comment]
Released in v1.2.0 :)

**Comment by anulman** [comment]
Woohoo! Thank you!

Meanwhile, I unfortunately caught a bug in this this weekend and was going to try and fix it before it landed 😅 see https://github.com/sgwilym/windups/pull/47

**Pull request merged**